cache/test_cache_control.py

```diff
@@ -0,0 +1,612 @@
+                    "{} header is present in the response".format(name))
+
+    def check_cached_response_headers(self, response):
+        for name, val in self.cached_headers.iteritems():
+            actual_val = response.headers.get(name, None)
+            if actual_val is None:
+                self.assertIsNone(actual_val,
+                    "{} header is missing in the cached response". \
+                        format(name))
+            else:
+                self.assertIsNotNone(actual_val,
+                    "{} header is present in the cached response". \
+                        format(name))
+
+    def _test(self):
+        self.start_all()
+        client = self.get_client('deproxy')
+        srv = self.get_server('deproxy')
+
+        req_headers = ''
+        if self.request_headers:
+            for name, val in self.request_headers.iteritems():
+                req_headers += '%s: %s\r\n' % (name, '' if val is None else val)
+        req = ("GET / HTTP/1.1\r\n"
+               "Host: localhost\r\n"
+               "%s\r\n" % req_headers)
+
+        response = self.client_send_req(client, req)
+        self.assertEqual(response.status, self.response_status,
+                         "request failed: {}, expected {}" \
+                         .format(response.status, self.response_status))
+        self.check_response_headers(response)
+
+        if self.sleep_interval:
+            time.sleep(self.sleep_interval)
+
+        cached_response = self.client_send_req(client, req)
+        self.assertEqual(response.status, self.cached_status,
+                         "request for cache failed: {}, expected {}" \
+                         .format(response.status, self.cached_status))
+
+        self.assertEqual(2, len(client.responses), "response lost")
+        if self.should_cache:
+            self.assertEqual(1, len(srv.requests),
+                             "response not cached as expected")
+        else:
+            self.assertEqual(2, len(srv.requests),
+                             "response is cached while it should not")
+ 
+        if self.should_cache:
+            self.check_cached_response_headers(cached_response)
+        else:
+            self.check_response_headers(cached_response)
+
+#########################################################
+#  cache_resp_hdr_del
+class CacheHdrDelBypass(TestCacheControl, tester.SingleTest):
+    tempesta_config = '''
+        cache_bypass * *;
+        cache_resp_hdr_del set-cookie Remove-me-2;
+        '''
+    response_headers = {'Set-Cookie': 'cookie=2; a=b', 'Remove-me-2': ''}
+    should_cache = False
+
+class CacheHdrDelFulfill(TestCacheControl, tester.SingleTest):
+    tempesta_config = '''
+        cache_fulfill * *;
+        cache_resp_hdr_del set-cookie Remove-me-2;
+        '''
+    response_headers = {'Set-Cookie': 'cookie=2; a=b', 'Remove-me-2': ''}
+    cached_headers = {'Set-Cookie': None, 'Remove-me-2': None}
+    should_cache = True
+
```

Comment:
@krizhanovsky Here is a question on feature under test rather than on test scenario.
With [#530](https://github.com/tempesta-tech/tempesta/issues/530) the header under `cache_resp_hdr_del` control must not appear in the response from the cache. However should it appear in the 1st response, forwarded from the backend?

- YES: in this case the response from cache differs from response from response from backemd. This violates the "cache sematic equivalence" requirement. Well, RFC 7234 omit this requirement (unlike previous HTTP specs). However I wonder if this is what you want.
- NO: in this case test scenario should be different and moreover, the implementation of the feature should be different.

Reply:
I haven't heard about any "cache semantic equivalence". For example, RFC 7234 5.2.2.2 clearly states:
```
any header fields in the response that have the field-name(s) listed MUST NOT
be sent in the response to a subsequent request without successfull revalidation
with the origin server.
```
So it's telling to remove only specific headers and only in subsequent requests.

Reply:
We need the option to not to store in the cache security sensitive way. Consider that a user retrieves a store cart within their session. The store cart may also contain some CSS, images and other static data, which is not user-specific (as for example the generated HTML with the goods). However, if the application isn't tuned too carefully, the server may send set-cooke with the static data. So we do want to store the static data in the cache to let other users use the data. However, we need to remove all sensitive data from the stored cache entity. In this case we forward the original response to the user, who generated the response, but all following requests should fetch the modified versions.

Reply:
Yes, now I understand the goal and it seems, everything is OK in this feature and it's tests.